File: src/utils/speech_detector.py
import sounddevice as sd
from scipy.io.wavfile import write
from collections import deque
import torch
import numpy as np

# CRITICAL: Use 16kHz for Silero VAD
sample_rate = 16000         # 16 kHz - Silero VAD requirement
chunk_size = 512            # number of samples for each VAD call (~32ms)
frame_ms = (chunk_size / sample_rate) * 1000.0  # ~32 ms

# Pre-roll amount in ms (how much audio we include before "start" is triggered)
# For example, 100 ms → about 3 frames of 32 ms each
pre_roll_ms = 100
num_pre_roll_frames = int(pre_roll_ms // frame_ms)  # ~3 frames

# Initialize VAD
print("Loading Silero VAD model...")
model_vad, utils = torch.hub.load(
    repo_or_dir='snakers4/silero-vad',
    model='silero_vad',
    force_reload=False,
    trust_repo=True
)
(get_speech_timestamps, save_audio, read_audio, VADIterator, collect_chunks) = utils
print("VAD model loaded successfully")

    
def record_audio(save_path: str = "recorded_audio.wav"):
    
    frames = []
    recording = False
    proper_start_sent = False
    print("Listening for speech...")
    print(f"Sample rate: {sample_rate} Hz, Chunk size: {chunk_size} samples (~{frame_ms:.1f} ms)")
    vad_iterator = VADIterator(model_vad)

    # Buffers for speech segment detection
    audio_buffer = bytearray()  # accumulate raw PCM for an utterance
    triggered = False

    # This ring buffer will store recent *silent* (non-triggered) chunks.
    # We’ll prepend these once we detect a "start" event.
    ring_buffer = deque(maxlen=num_pre_roll_frames)

    # We accumulate smaller incoming frames here until we reach `chunk_size` samples
    vad_buffer = np.array([], dtype=np.float32)
    
    try:
        while True:
            # CRITICAL: Record at 16kHz mono for Silero VAD
            data = sd.rec(int(5 * sample_rate), samplerate=sample_rate, channels=1, dtype='int16')
            sd.wait()  # Wait until recording is finished
            
            if not data.any():
                break

            # data is already a numpy array of shape (samples, 1), extract mono channel
            pcm_samples = data[:, 0]
            
            # Convert to float32 in range [-1, 1]
            audio_float32 = pcm_samples.astype(np.float32) / 32768.0
            vad_buffer = np.concatenate((vad_buffer, audio_float32))

            while len(vad_buffer) >= chunk_size:
                current_chunk = vad_buffer[:chunk_size]
                vad_buffer = vad_buffer[chunk_size:]
                
                # Call VAD iterator with correct sample rate
                speech_segments = vad_iterator(current_chunk, return_seconds=False)
                
                is_speech_start = (speech_segments is not None and 'start' in speech_segments)
                is_speech_end   = (speech_segments is not None and 'end'   in speech_segments)

                chunk_int16 = (current_chunk * 32768.0).astype(np.int16).tobytes()

                if is_speech_start and not triggered:
                    print(f"🎤 Speech START detected at frame {speech_segments['start']}")
                    # Prepend pre-roll frames
                    for rb_chunk in ring_buffer:
                        audio_buffer.extend(rb_chunk)
                    ring_buffer.clear()
                    triggered = True
                    proper_start_sent = False  # Reset for new utterance

                if triggered:
                    audio_buffer.extend(chunk_int16)
                    # Check if buffer meets threshold and proper start not sent
                    if len(audio_buffer) >= 24000 and not proper_start_sent:
                        print("✅ Proper speech start detected (>=0.75s)")
                        message_to_send = {'detection': 'proper_speech_start'}
                        # await websocket.send_text(str(message_to_send))
                        proper_start_sent = True
                else:
                    # Store in ring buffer for pre-roll
                    ring_buffer.append(chunk_int16)

                if is_speech_end and triggered:
                    print(f"🛑 Speech END detected at frame {speech_segments['end']}")
                    triggered = False

                    if len(audio_buffer) >= 24000:
                        # Save the recorded audio
                        audio_array = np.frombuffer(audio_buffer, dtype=np.int16)
                        write('recorded_audio.wav', sample_rate, audio_array)
                        print(f"💾 Audio saved: {len(audio_array)/sample_rate:.2f}s")
                        
                        # Clear buffer for next utterance
                        audio_buffer = bytearray()
                        proper_start_sent = False
                    return 0
    except KeyboardInterrupt:
        print("\n⏹️  Recording stopped by user")
        return frames

             
if __name__ == "__main__":
    audio_frames = record_audio()

File: kotlin/utils/SpeechDetector.kt
import javax.sound.sampled.*
import java.io.*
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*
import kotlin.collections.ArrayDeque
import kotlin.math.*

// Constants matching Python version
const val SAMPLE_RATE = 16000
const val CHUNK_SIZE = 512
const val FRAME_MS = (CHUNK_SIZE.toDouble() / SAMPLE_RATE) * 1000.0  // ~32 ms
const val PRE_ROLL_MS = 100
const val NUM_PRE_ROLL_FRAMES = (PRE_ROLL_MS / FRAME_MS).toInt()  // ~3 frames

class SpeechDetector {
    private var recording = false
    private var properStartSent = false
    private val frames = mutableListOf<ByteArray>()
    
    // Buffers for speech segment detection
    private val audioBuffer = ByteArrayOutputStream()
    private var triggered = false
    
    // Ring buffer for pre-roll
    private val ringBuffer = ArrayDeque<ByteArray>(NUM_PRE_ROLL_FRAMES)
    
    // VAD buffer
    private val vadBuffer = mutableListOf<Float>()
    
    // Simple VAD implementation (placeholder for Silero VAD)
    private var silenceCounter = 0
    private var speechCounter = 0
    
    init {
        println("Speech Detector initialized")
        println("Sample rate: $SAMPLE_RATE Hz, Chunk size: $CHUNK_SIZE samples (~${FRAME_MS.toInt()} ms)")
    }
    
    fun recordAudio(savePath: String = "recorded_audio.wav"): Int {
        println("Listening for speech...")
        
        val audioFormat = AudioFormat(
            AudioFormat.Encoding.PCM_SIGNED,
            SAMPLE_RATE.toFloat(),
            16, // 16-bit
            1,  // mono
            2,  // frame size: 1 channel * 2 bytes per sample
            SAMPLE_RATE.toFloat(),
            false // little endian
        )
        
        val info = DataLine.Info(TargetDataLine::class.java, audioFormat)
        
        if (!AudioSystem.isLineSupported(info)) {
            println("Audio recording not supported!")
            return -1
        }
        
        val line = AudioSystem.getLine(info) as TargetDataLine
        
        try {
            line.open(audioFormat)
            line.start()
            println("🎤 Recording started...")
            
            val buffer = ByteArray(CHUNK_SIZE * 2) // 2 bytes per sample
            
            while (true) {
                val bytesRead = line.read(buffer, 0, buffer.size)
                
                if (bytesRead > 0) {
                    val end_recording = processAudioChunk(buffer, bytesRead, savePath)
                    // Check if we should stop recording
                    if (end_recording) {
                        // Recording complete
                        break
                    }
                }
                
                // Check for user interrupt (simple timeout for demo)
                if (silenceCounter > 1000) { // ~32 seconds of silence
                    println("Recording stopped due to long silence")
                    break
                }
            }
            
        } catch (e: Exception) {
            println("Recording error: ${e.message}")
            return -1
        } finally {
            line.stop()
            line.close()
        }
        
        return 0
    }
    
    private fun processAudioChunk(buffer: ByteArray, bytesRead: Int, savePath: String): Boolean {
        // Convert bytes to float array for VAD processing
        val samples = bytesRead / 2 // 2 bytes per sample
        val audioFloat = FloatArray(samples)
        
        val byteBuffer = ByteBuffer.wrap(buffer, 0, bytesRead).order(ByteOrder.LITTLE_ENDIAN)
        for (i in 0 until samples) {
            val sample = byteBuffer.short
            audioFloat[i] = sample.toFloat() / 32768.0f
        }
        
        vadBuffer.addAll(audioFloat.toList())
        
        while (vadBuffer.size >= CHUNK_SIZE) {
            val currentChunk = vadBuffer.take(CHUNK_SIZE).toFloatArray()
            repeat(CHUNK_SIZE) { vadBuffer.removeFirst() }
            
            // Simple VAD (replace with actual Silero VAD if needed)
            val (isSpeechStart, isSpeechEnd) = simpleVAD(currentChunk)
            
            val chunkBytes = floatArrayToBytes(currentChunk)
            
            if (isSpeechStart && !triggered) {
                println("🎤 Speech START detected")
                
                // Prepend pre-roll frames
                for (rbChunk in ringBuffer) {
                    audioBuffer.write(rbChunk)
                }
                ringBuffer.clear()
                triggered = true
                properStartSent = false
            }
            
            if (triggered) {
                audioBuffer.write(chunkBytes)
                
                // Check if buffer meets threshold and proper start not sent
                if (audioBuffer.size() >= 24000 && !properStartSent) {
                    println("✅ Proper speech start detected (>=0.75s)")
                    properStartSent = true
                }
            } else {
                // Store in ring buffer for pre-roll
                if (ringBuffer.size >= NUM_PRE_ROLL_FRAMES) {
                    ringBuffer.removeFirst()
                }
                ringBuffer.addLast(chunkBytes)
            }
            
            if (isSpeechEnd && triggered) {
                println("🛑 Speech END detected")
                triggered = false
                
                if (audioBuffer.size() >= 24000) {
                    // Save the recorded audio
                    saveAudioToFile(audioBuffer.toByteArray(), savePath)
                    val duration = audioBuffer.size().toDouble() / (SAMPLE_RATE * 2)
                    println("💾 Audio saved: $duration")

                    // Clear buffer for next utterance
                    audioBuffer.reset()
                    properStartSent = false
                    return true
                }
            }
        
        }
        return false
    }
    
    private fun simpleVAD(audioChunk: FloatArray): Pair<Boolean, Boolean> {
        // Simple energy-based VAD (replace with Silero VAD for production)
        val energy = audioChunk.map { it * it }.average()
        val threshold = 0.001 // Adjust based on your environment
        
        val isSpeech = energy > threshold
        
        var speechStart = false
        var speechEnd = false
        
        if (isSpeech) {
            speechCounter++
            silenceCounter = 0
            
            // Start speech if we have enough consecutive speech frames
            if (speechCounter == 3 && !triggered) {
                speechStart = true
            }
        } else {
            silenceCounter++
            speechCounter = 0
            
            // End speech if we have enough consecutive silence frames
            if (silenceCounter == 10 && triggered) {
                speechEnd = true
            }
        }
        
        return Pair(speechStart, speechEnd)
    }
    
    private fun floatArrayToBytes(floatArray: FloatArray): ByteArray {
        val byteArray = ByteArray(floatArray.size * 2)
        val byteBuffer = ByteBuffer.wrap(byteArray).order(ByteOrder.LITTLE_ENDIAN)
        
        for (sample in floatArray) {
            val intSample = (sample * 32767).toInt().coerceIn(-32768, 32767)
            byteBuffer.putShort(intSample.toShort())
        }
        
        return byteArray
    }
    
    private fun saveAudioToFile(audioData: ByteArray, filename: String) {
        try {
            val audioFormat = AudioFormat(
                AudioFormat.Encoding.PCM_SIGNED,
                SAMPLE_RATE.toFloat(),
                16, // 16-bit
                1,  // mono
                2,  // frame size
                SAMPLE_RATE.toFloat(),
                false // little endian
            )
            
            val audioInputStream = AudioInputStream(
                ByteArrayInputStream(audioData),
                audioFormat,
                audioData.size / audioFormat.frameSize.toLong()
            )
            
            val file = File(filename)
            AudioSystem.write(audioInputStream, AudioFileFormat.Type.WAVE, file)
            println("Audio saved to: ${file.absolutePath}")
            
        } catch (e: Exception) {
            println("Error saving audio: ${e.message}")
        }
    }
}

fun main() {
    println("=== Speech Detector Test ===")
    
    val detector = SpeechDetector()
    
    println("Starting speech detection...")
    println("Speak into your microphone. The detector will:")
    println("- Start recording when speech is detected")
    println("- Stop recording after speech ends")
    println("- Save audio to 'recorded_audio.wav'")
    println("\nPress Ctrl+C to stop")
    
    try {
        val result = detector.recordAudio("test_recording.wav")
        
        if (result == 0) {
            println("Recording completed successfully!")
        } else {
            println("Recording failed with code: $result")
        }
        
    } catch (e: Exception) {
        println("Error during recording: ${e.message}")
    }
    
    println("Speech detection test complete!")
}
